brave/utility/tor/tor_impl.cc

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2018 The Brave Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "brave/utility/tor/tor_impl.h"
+
+#include <utility>
+
+#include "base/command_line.h"
+#include "base/process/kill.h"
+#include "base/process/launch.h"
+#include "base/single_thread_task_runner.h"
+#include "base/task_scheduler/post_task.h"
+#include "base/threading/thread.h"
+#include "base/threading/thread_task_runner_handle.h"
+
+#if defined(OS_POSIX)
+int pipehack[2];
+
+static void SIGCHLDHandler(int signal) {
+  pid_t p;
+  int status;
+
+  if ((p = waitpid(-1, &status, WNOHANG | WUNTRACED)) != -1) {
+    if (WIFSIGNALED(status) || WCOREDUMP(status)) {
+      write(pipehack[1], &p, sizeof(pid_t));
+    } else if (WIFEXITED(status)) {
+      // TODO(darkdh): Handle "Address already in use" issue. Not notifying as
+      // workaroud for now to prevent infinitely relaunch
+      if (!WEXITSTATUS(status))
+        write(pipehack[1], &p, sizeof(pid_t));
+    }
+  }
+}
+
+static void SetupPipeHack() {
+  if (pipe(pipehack) == -1)
+    LOG(ERROR) << "pipehack";
+
+  int flags;
+  for (size_t i = 0; i < 2; ++i) {
+    if ((flags = fcntl(pipehack[i], F_GETFL)) == -1)
+      LOG(ERROR) << "get flags";
+    if (i == 1)
+      flags |= O_NONBLOCK;
+    flags |= O_CLOEXEC;
+    if (fcntl(pipehack[i], F_SETFL, flags) == -1)
+      LOG(ERROR) << "set flags";
+  }
+
+  struct sigaction action;
+  memset(&action, 0, sizeof(action));
+  action.sa_handler = SIG_IGN;
+  sigaction(SIGPIPE, &action, NULL);
+
+  memset(&action, 0, sizeof(action));
+  action.sa_handler = SIGCHLDHandler;
+  sigaction(SIGCHLD, &action, NULL);
+}
+#endif
+
+namespace brave {
+
+class TorLauncherDelegate : public base::LaunchOptions::PreExecDelegate {
+ public:
+    TorLauncherDelegate() {}
+    ~TorLauncherDelegate() override {}
+
+    void RunAsyncSafe() override {
+#if defined(OS_POSIX)
+      for (size_t i = 0; i < 2; ++i)
+        close(pipehack[i]);
+#endif
+      // TODO(darkdh): doesn't prevent ctrl+c propagated by terminal
```

Comment:
Can you expand on this?

Reply:
When `npm start` launching Brave with tor process launched, you ctrl+c on the terminal running `npm`. Tor process will get terminated before TorImpl destructor terminates it.
This behavior will trigger SIGCHLD handler and do a relaunch (current behavior) but it will be terminated by `~TorImpl` a moment later. 

Setting ignore SIGINT doesn't help to mitigate this case, not a big deal because tor process will eventually be terminated. Just wanna make it looks better

Reply:
fixed in fea7dca